### ISSUE TITLE
articulations_points returns repeated nodes

### ISSUE THREAD

**Issue by malbarbo** [networkx/networkx#2331]
```python
import networkx as nx
from networkx.algorithms.components.biconnected import articulation_points

g = nx.Graph()
g.add_edges_from([(0, 1), (1, 2), (1, 3)])
assert list(articulation_points(g)) == [1, 1]
```

**Comment by leotrs** [comment]
After investigating the problem, node `1` is returned twice because the algorithm identifies two different biconnected components that split from the original graph after removing `1`. This is due to the fact that the algorithm proceeds by DFS traversal. Starting at `0`, node `1` will be identified as articulation point after traversing down its child `2` (and everything that follows from `2`), and once again after traversing down node `3`.

It seems to me that this behavior is already known, since [all tests](https://github.com/networkx/networkx/blob/6e20b952a957af820990f68d9237609198088816/networkx/algorithms/components/tests/test_biconnected.py) that involve `articulation_points` call `set` over the returned list (thus deleting the duplicate values).

@malbarbo I would advice just calling `set` over the return value, as it seems that there is some valuable information in the repeated values in the returned list of `articulation_points`.

@jfinkels On the other hand, if this is not expected behavior, it should be easy to add a quick memoification step inside `articulation_points` so that the same node is not yielded more than once. I could do this if there is interest.

**Comment by jfinkels** [comment]
It seems the function currently yields the node once each time it is discovered to lie in the intersection of two maximal biconnected subgraphs. In this example, the node 1 is in the intersection of the two maximal biconnected components {0, 1} and {1, 2} and in the intersection of {0, 1} and {1, 3}. This is a little confusing if we just consider the name of the function, `articulation_points`, which seems to me to imply that it yields each articulation point once.

My feeling is that this "multiplicity" information about articulation points could be expressed more clearly in a different way (see the [Related structures][1] section of the Wikipedia article on biconnected components for some possibilities). Therefore, I think that this function should yield each articulation point only once:

```python
seen = set()
for v in _biconnected_dfs(G, components=False):
    if v not in seen:
        yield v
```

A test should also be added to check that node are yielded only once (the example code by @malbarbo is fine).

[1]: https://en.wikipedia.org/wiki/Biconnected_component#Related_structures

**Comment by leotrs** [comment]
The proposed PR should take care of this.

About this "multiplicity" information. This needs confirmation but my working hypothesis is that each articulation point appears in `list(_biconnected_dfs(G))` a number of times equal to the increase in the number of components in the network after removing that articulation point. In @malbarbo's example, `1` appeared twice because the network would break up in three components after removing `1`, thus increasing the number of components by 2.

If this is the case, then we could implement a new function, say
```python
def number_of_components_articulated_at(G):
    count = Counter(list(_biconnected_dfs(G)))
    return {node: count[node] + 1 for node in count.keys()}
```

**Comment by malbarbo** [comment]
The fix is great! Thanks.

**Comment by leotrs** [comment]
Solved by #2333.

@jfinkels thoughts on the proposed `number_of_components_articulated_at`?

**Comment by jfinkels** [comment]
What I was imagining was creating a function for computing either the "block graph" or "block-cut tree" as described in the Wikipedia article to which I linked above. From these structures, the number of blocks in which each articulation point participates can be read off more easily. The length of the name of the function you are suggesting is an indicates to me that it may be too specific :)

**Comment by hagberg** [comment]
Fixed by #2333.
Open a new issue for further enhancements.
